connexion/decorators/response.py

```diff
@@ -0,0 +1,190 @@
+import abc
+import collections.abc
+import functools
+import logging
+import types
+import typing as t
+from enum import Enum
+
+from connexion.datastructures import NoContent
+from connexion.exceptions import NonConformingResponseHeaders
+from connexion.frameworks.abstract import Framework
+from connexion.lifecycle import ConnexionResponse, MiddlewareResponse
+from connexion.operations import AbstractOperation
+from connexion.utils import is_json_mimetype
+
+logger = logging.getLogger(__name__)
+
+
+class BaseResponseDecorator:
+    def __init__(
+        self, operation: AbstractOperation, *, framework: t.Type[Framework], jsonifier
+    ):
+        self.operation = operation
+        self.framework = framework
+        self.jsonifier = jsonifier
+
+    @abc.abstractmethod
+    def __call__(self, function: t.Callable) -> t.Callable:
+        raise NotImplementedError
+
+    def build_framework_response(self, handler_response):
+        data, status_code, headers = self._unpack_handler_response(handler_response)
+        content_type = self._deduct_content_type(data, headers)
+        if not self.framework.is_framework_response(data):
+            data, status_code = self._prepare_body_and_status_code(
+                data, status_code=status_code, mimetype=content_type
+            )
+        return self.framework.build_response(
+            data, content_type=content_type, status_code=status_code, headers=headers
+        )
+
+    def _deduct_content_type(self, data: t.Any, headers: dict) -> str:
+        """Deduct the response content type from the returned data, headers and operation spec.
+
+        :param data: Response data
+        :param headers: Headers returned by the handler.
+
+        :return: Deducted content type
+
+        :raises: NonConformingResponseHeaders if content type cannot be deducted.
+        """
+        content_type = headers.get("Content-Type")
+
+        # TODO: don't default
+        produces = list(set(self.operation.produces))
+        if data is not None and not produces:
```

Comment:
This keeps backward compatibility, although I don't think we should. We should expect the spec to define a content type when a response is returned.

https://swagger.io/specification/v2/#responseObject
- A missing response schema defines an empty response
- If a response schema is present, produces should be defined

In OpenAPI 3, a response cannot be defined without its content type.